src/raftstore/store/snap.rs

```diff
@@ -1302,15 +1302,19 @@ impl SnapManager {
         self.max_total_size
     }
 
-    pub fn register(&self, key: SnapKey, entry: SnapEntry) {
+    // For `SnapEntry::Sending` we need to support register multi times because
+    // Raft can send same snapshot to many peers (e.g. a follower and several learners).
+    // But for others, Only support register one time.
+    // TODO: make it work as a state machine, and change `Vec<SnapEntry>` to `SnapEntry`.
+    pub fn register(&self, key: SnapKey, entry: SnapEntry) -> Result<()> {
         debug!("register [key: {}, entry: {:?}]", key, entry);
-        let mut core = self.core.wl();
-        match core.registry.entry(key) {
+        match self.core.wl().registry.entry(key) {
```

Comment:
I remember that using `wl` in the match may cause deadlock, so @overvenus moved it outside before for some locks.

Any reason to use in the match again?

Reply:
The lock won't be released before any branch of match finished. But here it's just what we want: we must keep the lock to update the map.

Reply:
no, seem the lifetime will be longer than match. @overvenus

Reply:
/cc @overvenus

Reply:
I think it is better to move it out, because it looks like the lock is released after the `match ... {` but actually it is not. Move it out, so we know for sure the lock is not released.